torch_np/_funcs_impl.py

```diff
@@ -1208,6 +1228,68 @@ def outer(a: ArrayLike, b: ArrayLike, out: Optional[OutArray] = None):
     return torch.outer(a, b)
 
 
+def einsum(*operands, out=None, optimize=False, **kwargs):
+    # Have to manually normalize *operands and **kwargs, following the NumPy signature
+    # >>> np.einsum?
+    # Signature: np.einsum(*operands, out=None, optimize=False, **kwargs)
+    # Docstring:
+    # einsum(subscripts, *operands, out=None, dtype=None, order='K',
+    #       casting='safe', optimize=False)
+
+    from ._normalizations import (
+        maybe_copy_to,
+        normalize_casting,
+        normalize_dtype,
+        normalize_not_implemented,
+        normalize_outarray,
+        wrap_tensors,
+    )
+
+    dtype = normalize_dtype(kwargs.pop("dtype", None))
+    casting = normalize_casting(kwargs.pop("casting", "safe"))
+
+    parm = lambda _: None  # a fake duck-typed inspect.Parameter stub
+    parm.name = "out"
+    out = normalize_outarray(out, parm=parm)
+
+    parm.default = "K"
+    parm.name = "order"
+    order = normalize_not_implemented(kwargs.pop("order", "K"), parm=parm)
+    if kwargs:
+        raise TypeError("unknown arguments: ", kwargs)
+
+    # parse arrays and normalize them
+    if isinstance(operands[0], str):
+        # ("ij->", arrays) format
+        sublist_format = False
+        subscripts, array_operands = operands[0], operands[1:]
+    else:
+        # op, str, op, str ... format: normalize every other argument
+        sublist_format = True
+        array_operands = operands[:-1][::2]
+
+    tensors = [normalize_array_like(op) for op in array_operands]
+    target_dtype = (
+        _dtypes_impl.result_type_impl([op.dtype for op in tensors])
+        if dtype is None
+        else dtype
+    )
+    tensors = _util.typecast_tensors(tensors, target_dtype, casting)
+
+    if sublist_format:
+        # recombine operands
+        sublists = operands[1::2]
+        sublistout = (operands[-1],) if len(operands) % 2 == 1 else ()
+        operands = builtins.sum((_ for _ in zip(tensors, sublists)), ()) + sublistout
```

Comment:
It took me a bit of thinking to understand what was going on here. What about the more explicit using `chain`
```suggestion
        operands = list(itertools.chain(*zip(tensor, operands[1::2])))
        if len(operands) % 2 == 1:
            operands.append(operands[-1])
```

Reply:
Have to admit I kind of liked the _what the ..._ vibe of the `sum(..., ())` trick :-). Anyway, changed with some small tweaks for readability.